### PR TITLE
fix: Update NonMatchingSelector tests and fixed failing TSOwnable.t.sol file

### PR DESCRIPTION
**Pull request by AmadiMichael**
Great work here guys! just wanted to contribute to fixes to my findings

```solidity
/// @notice Test that a non-matching selector reverts
  function testNonMatchingSelector(bytes32 callData) public {
    bytes8[] memory func_selectors = new bytes8[](2);
    func_selectors[0] = bytes8(hex"13af4035");
    func_selectors[1] = bytes8(hex"8da5cb5b");

    bytes8 func_selector = bytes8(callData >> 0xe0);
    for (uint256 i = 0; i < 2; i++) {
      if (func_selector != func_selectors[i]) {
        return;
      }
    }

    address target = address(owner);
    uint256 OneWord = 0x20;
    bool success = false;
    assembly {
      success := staticcall(
          gas(),
          target,
          add(callData, OneWord),
          mload(callData),
          0,
          0
      )
    }
    assert(!success);
  }
```

This test found in most auth tests is bound to always succeed without really testing it, actually there's no chance a call to this function would reach the assembly part.

Example: if the fuzzer parses in `0xb10e2d527612073b26eecdfd717e6a320cf44b4afac2b0732d9fcbe2b7fa0cf6`, the array is filled up with the function sigs the contract has but then, firstly this line `bytes8 func_selector = bytes8(callData >> 0xe0);` is bound to return `0x0000000000000000`. Using the bytes32 above this will firstly execute (0xb10e2d527612073b26eecdfd717e6a320cf44b4afac2b0732d9fcbe2b7fa0cf6 >> 0xe0) which is `0xb10e2d52` with 224 bits of zero to the left. Performing bytes8(0xb10e2d52) will only take the first 8 bytes of this which is all zeroes.

I fixed it in assembly with 
```solidity
bytes4 func_selector;
        assembly {
            func_selector := and(
                callData,
                0xffffffff00000000000000000000000000000000000000000000000000000000
            )
        }
```

Secondly, even if the above issue didn't exist, the if statement states that ` if (func_selector != func_selectors[i]) {
        return;
      }` while looping through the array of signatures. The issue is that rather than checking that if the shifted bytes above returns the same bytes as a function sig in the array it should return as that's not the aim of the test but rather it returns if the shifted bytes is different from any one in the array meaning that it would only get to the assembly part if the sig is the same and which will obviously pass. The fix was to change `!=` to `==`

The third issue was the assembly part. It wasn't detected i believe because the test never got there but if it did the execution would have ran out of gas. In the call it asks the evm to read from `add(callData, OneWord)` which could be as much as uint256.max + 0x80 also it asks it to read up to the latter plus mload(calldata) which could be doubling the size minus 0x80. 

The fix was this 
```solidity
bool success = false;
        assembly {
            mstore(0x80, callData)
            // if its a state changing call this will return 0
            success := staticcall(gas(), target, 0x80, 0x20, 0, 0)

            if iszero(success) {
                success := call(gas(), target, 0, 0x80, 0x20, 0, 0)
            }
        }
```

Checking a static call first incase which if it fails we make a call. mstoring calldata in memory and using it.

The second was the TSOwnable failing test (without a need to add config_with_create_2), by separating the HuffConfig deployment and storing the returned address.

### CODE REVIEW

File: test/test-utils/NonMatchingSelectorHelper.sol [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * expected to fail
+ */
+abstract contract NonMatchingSelectorsHelper {
+    function nonMatchingSelectorHelper(
+        bytes4[] memory func_selectors,
+        bytes32 callData,
+        address target
+    ) internal returns (bool) {
+        bytes4 func_selector;
```
Comment (MathisGD) [inline test/test-utils/NonMatchingSelectorHelper.sol:15]: Don't this work ?
```suggestion
        bytes4 func_selector = bytes4(calldata); 
```

File: test/auth/Auth.t.sol [hunk 2]
```diff
@@ -53,36 +54,21 @@ contract AuthTest is Test {
...
-    for (uint256 i = 0; i < 6; i++) {
-      if (func_selector != func_selectors[i]) {
-        return;
-      }
+    function testNonMatchingSelector(bytes32 callData) public {
+        bytes4[] memory func_selectors = new bytes4[](6);
+        func_selectors[0] = bytes4(hex"13af4035");
+        func_selectors[1] = bytes4(hex"7a9e5e4b");
+        func_selectors[2] = bytes4(hex"8da5cb5b");
+        func_selectors[3] = bytes4(hex"bf7e214f");
```
Comment (MathisGD) [inline test/auth/Auth.t.sol:62]:
```suggestion
        func_selectors[0] = Auth.setOwner.selector;
        func_selectors[1] = Auth.setAuthority.selector;
        func_selectors[2] = Auth.owner.selector;
        func_selectors[3] = Auth.authority.selector;
```

File: test/auth/Owned.t.sol [hunk 3]
```diff
@@ -26,33 +28,23 @@ contract OwnedTest is Test {
...
 
   /// @notice Test that a non-matching selector reverts
-  function testNonMatchingSelector(bytes32 callData) public {
-    bytes8[] memory func_selectors = new bytes8[](2);
-    func_selectors[0] = bytes8(hex"13af4035");
-    func_selectors[1] = bytes8(hex"8da5cb5b");
+    function testNonMatchingSelector(bytes32 callData) public {
+        bytes4[] memory func_selectors = new bytes4[](2);
+        func_selectors[0] = bytes4(hex"13af4035");
+        func_selectors[1] = bytes4(hex"8da5cb5b");
```
Comment (MathisGD) [inline test/auth/Owned.t.sol:34]:
```suggestion
        func_selectors[0] = Owned.setOwner.selector;
        func_selectors[1] = Owned.owner.selector;
```

File: test/auth/Owned.t.sol [hunk 4]
```diff
@@ -26,33 +28,23 @@ contract OwnedTest is Test {
...
-      success := staticcall(
-          gas(),
-          target,
-          add(callData, OneWord),
-          mload(callData),
-          0,
-          0
-      )
+        // bytes4 func_selector = bytes4(callData << 0xe0);
+        // the above will always return 0 because after shifting all left bits are 00000000
```
Comment (MathisGD) [inline test/auth/Owned.t.sol:39]: I would remove if not useful in the current implementation

File: test/auth/RolesAuthority.t.sol [hunk 5]
```diff
@@ -41,36 +43,19 @@ contract RolesAuthorityTest is Test {
...
-          add(callData, OneWord),
-          mload(callData),
-          0,
-          0
-      )
+  /// @notice Test that a non-matching selector reverts
+    function testNonMatchingSelector(bytes32 callData) public {
+        bytes4[] memory func_selectors = new bytes4[](2);
+        func_selectors[0] = bytes4(hex"13af4035");
+        func_selectors[1] = bytes4(hex"8da5cb5b");
```
Comment (MathisGD) [inline test/auth/RolesAuthority.t.sol:50]: Doesn't the contract have 6 selectors ?

**Review by MathisGD — COMMENTED** [review]
Great finding ! A few comments :

File: test/test-utils/NonMatchingSelectorHelper.sol (hunk 1 above)
Comment (AmadiMichael) [reply test/test-utils/NonMatchingSelectorHelper.sol:15]: Ah yes, totally

File: test/utils/TSOwnable.t.sol [hunk 6]
```diff
@@ -13,16 +13,16 @@ interface TSOwnable {
...
+        huffConfig = address(HuffDeployer.config());
         tsOwnable = TSOwnable(
-            HuffDeployer
-                .config()
-                .with_code(wrapper_code)
-                .deploy("utils/TSOwnable")
+            HuffConfig(huffConfig).with_code(wrapper_code).deploy(
+                "utils/TSOwnable"
+            )
         );
```
[the remark this answers is not in the document]
Comment (AmadiMichael) [reply test/utils/TSOwnable.t.sol:26]: Alright, do you suggest i leave it as it is or any changes?

File: test/auth/RolesAuthority.t.sol (hunk 5 above)
Comment (AmadiMichael) [reply test/auth/RolesAuthority.t.sol]: Yes, added them in the latest commit

File: test/utils/TSOwnable.t.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (MathisGD) [reply test/utils/TSOwnable.t.sol:26]: Idk what's the strategy for this kind of PRs. @cheethas ?

**Review by MathisGD — APPROVED** [review]

File: test/utils/TSOwnable.t.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (cheethas) [reply test/utils/TSOwnable.t.sol:26]: Im not too fussed as they're the same approach, we can just rebase removing the commit to fix the conflict then get it merged in

**Review by cheethas — APPROVED** [review]
Nice, thank you both @MathisGD and @AmadiMichael

**Review by MathisGD — APPROVED** [review]

**Pull request merged**